sn_client/src/wallet.rs

```diff
@@ -104,49 +94,67 @@
+        let mut proofs = PaymentTransactionsMap::default();
+        let addrs_to_pay: Vec<NetworkAddress> = content_addrs
             .filter(|name| {
                 if let Some(proof) = self.wallet.get_payment_proof(name) {
-                    proofs.insert(**name, proof.clone());
+                    proofs.insert(name.clone(), proof.clone());
                     false
                 } else {
                     true
                 }
             })
             .collect();
 
-        let number_of_records_to_pay = addrs_to_pay.len() as u64;
+        let _number_of_records_to_pay = addrs_to_pay.len() as u64;
         // If no addresses need to be paid for, we don't have to go further
         if addrs_to_pay.is_empty() {
             trace!("We already hold payment proofs for all the Chunks.");
             return Ok((proofs, None));
         }
 
-        // Let's build the payment proofs for list of content addresses
-        let (root_hash, audit_trail_info) = build_payment_proofs(addrs_to_pay.into_iter())?;
-        let num_of_addrs = audit_trail_info.len() as u64;
+        let mut total_cost = Token::zero();
 
-        // Always check storage cost, and overpay to allow margin when validation.
-        self.set_store_cost_from_random_address().await?;
-        let storage_cost = Token::from_nano(
-            self.store_cost().as_nano() * (2.0f64.powf((num_of_addrs / 100 + 1) as f64)) as u64,
-        );
+        for content_addr in addrs_to_pay {
+            // TODO: parallelise this
+            let (payments, cost) = self
+                .pay_for_storage_at_address(&content_addr, verify_store)
+                .await?;
 
-        info!("Storage cost per record: {}", storage_cost);
+            if let Some(cost) = total_cost.checked_add(cost) {
+                total_cost = cost;
+            }
 
-        let amount_to_pay = number_of_records_to_pay * storage_cost.as_nano();
+            // store the payment proof
+            proofs.insert(content_addr.clone(), payments);
+        }
 
-        trace!("Making payment for {num_of_addrs} addresses of {amount_to_pay:?} nano tokens.");
+        Ok((proofs, Some(total_cost)))
+    }
 
+    /// Send tokens to nodes closest to the data we want to make storage payment for.
+    ///
+    /// Returns DbcId created for the payment, storage cost is for that record
+    ///
+    /// This can optionally verify the store has been successful (this will attempt to GET the dbc from the network)
+    pub async fn pay_for_storage_at_address(
+        &mut self,
+        content_addr: &NetworkAddress,
+        verify_store: bool,
+    ) -> Result<(Vec<DbcId>, Token)> {
+        let amount_to_pay = self.get_store_cost_at_address(content_addr).await?;
+
+        trace!("Making payment for of {amount_to_pay:?} nano tokens to each node in close group at {content_addr:?}.");
+
+        // TODO: This needs to go out to each CLOSEGROUP of addresses
```

Comment:
you mean paying each individual holder ?

hmm...  I don't think we need to contact each holder individually, 
just `get_closest(chunk_addr)` then transfer some amount to each fetched close_peer, and send the transfer to network shall be enough?

just, that means the payment_proof needs to hold 8 spend_ids, and on validation, the holder only needs to validate the spend to itself (or still ALL for security?)

Reply:
Each CLOSE_GROUP for each record address needs to return a price, which should then be paid. 

on validation, each would just do itself, I think?

Reply:
> Each CLOSE_GROUP for each record address needs to return a price, which should then be paid.

ok. I thought a close_group can charge with the same.

> on validation, each would just do itself, I think?

yeah. but that's can be done on receiving the record via `put_record` , instead of contacting them directly individually.